### PR TITLE
Enable 'edit specific price' button on Product page

### PR DESCRIPTION
**Pull request by matks**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | In BackOffice "Catalog" > "Products", currently admins can either "add" or "delete" a "specific price" on the "Edit a Product" page. This PR enables to edit a specific price.
| Type?         | new feature
| Category?     | BO
| BC breaks?    | yes, see below
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/FF-160
| How to test?  | See below

## Possible BC break

Currently, in the "Edit a product" page, **all the forms** are being loaded at once using `PrestaShopBundle\Form\Admin\Product\ProductPrice` Symfony Form.
But the "Add a Specific Price" form is submitted through an AJAX call.

I reused the submit ajax call however, with this PR merged, now the "edit a specific price" button removes the initial "Add a specific price" form and replaces it with another `PrestaShopBundle\Form\Admin\Product\EmbeddedSpecificPrice`. The behavior is iso to `ProductPrice` except is someone overrides it. The submit process has not changed, hooks are preserved.

## How to test

Go on a "Edit a product" page on the BackOffice. Add a specific price. It appears on the page. Now you can edit any of its properties (user group ? customer ? combination ? from and to dates ? how much ? percentage or fixed price ?) and check that the edit process works correctly.

Be careful: in multi-shop context another field appears. Also products with combination have another field that single-combination products do not display.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9388)

### CODE REVIEW

**Comment by mickaelandrieu** [comment]
Hello @matks, regarding the branch of your pull request only develop branch is concerned.

More, we don't do any BC break on 1.6.1.x branch, thanks for your understanding :+1:

**Comment by matks** [comment]
@mickaelandrieu Sorry this is a mistake about me adding labels

File: src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/typeahead.html.twig [hunk 1]
```diff
@@ -221,6 +232,7 @@
 
                 modalConfirmation.create(translate_javascripts['Are you sure to delete this?'], null, {
```
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/typeahead.html.twig:233]: Domains needed? If yes: _Admin.Notifications.Warning_

Comment (mickaelandrieu) [reply src/PrestaShopBundle/Resources/views/Admin/TwigTemplateForm/typeahead.html.twig:233]: Probably but not here (this is a super weirdo custom translation system, already translated somewhere with the right domain).

**Review by eternoendless — CHANGES_REQUESTED** [review]
I think that the form.js file should be moved to the new theme if possible.

> Reviewed 9 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9388#-:-LIl2j6t9MF3WwoNWvR5:brcfdsf)* status: all files reviewed, 13 unresolved discussions (waiting on @LouiseBonnard and @matks)

___
*[admin-dev/themes/default/js/bundle/product/form.js, line 621 at r1](https://reviewable.io/reviews/prestashop/prestashop/9388#-LIkKdUF4fwQs78IiQ11:-LIkKdUG8shXBBaiXmDh:b-yj3upb) ([raw file](https://github.com/prestashop/prestashop/blob/b0341caf1bafabd9bf3d50f96ce5226180863c26/admin-dev/themes/default/js/bundle/product/form.js#L621)):*
> ```JavaScript
> 
>             // setup listeners for "edit" action
>             $('.specific-price-edit').click(function () {
> ```

As this is done within a loop, won't it add multiple listeners on all buttons?
___
*[admin-dev/themes/default/js/bundle/product/form.js, line 739 at r1](https://reviewable.io/reviews/prestashop/prestashop/9388#-LIkLcpZ-wjBL9IpiN8t:-LIkLcpZ-wjBL9IpiN8u:bb5rtji) ([raw file](https://github.com/prestashop/prestashop/blob/b0341caf1bafabd9bf3d50f96ce5226180863c26/admin-dev/themes/default/js/bundle/product/form.js#L739)):*
> ```JavaScript
>         type: 'GET',
>         url: url,
>         success: function(specific_price_update_form) {
> ```

What happens if the request fails for some reason?
___
*[admin-dev/themes/default/js/bundle/product/form.js, line 739 at r1](https://reviewable.io/reviews/prestashop/prestashop/9388#-LIkLgb53SH-pHCvu2XE:-LIkLgb53SH-pHCvu2XF:b-td7kuk) ([raw file](https://github.com/prestashop/prestashop/blob/b0341caf1bafabd9bf3d50f96ce5226180863c26/admin-dev/themes/default/js/bundle/product/form.js#L739)):*
> ```JavaScript
>         type: 'GET',
>         url: url,
>         success: function(specific_price_update_form) {
> ```

Please use camelCase for names: https://github.com/airbnb/javascript#naming--camelCase
___
*[admin-dev/themes/default/js/bundle/product/form.js, line 742 at r1](https://reviewable.io/reviews/prestashop/prestashop/9388#-LIkLxWvAND3NvW4mySZ:-LIkLxWvAND3NvW4myS_:b-k7viwz) ([raw file](https://github.com/prestashop/prestashop/blob/b0341caf1bafabd9bf3d50f96ce5226180863c26/admin-dev/themes/default/js/bundle/product/form.js#L742)):*
> ```JavaScript
>           // remove "add a price" form and replace it with "edit a price"  loadedform
>           $('#specific_price_form').empty();
>           $('#specific_price_form').append(specific_price_update_form);
> ```

Please avoid duplicate selectors (store in a variable)
___
*[admin-dev/themes/default/js/bundle/product/form.js, line 755 at r1](https://reviewable.io/reviews/prestashop/prestashop/9388#-LIkM93t4QWP7p2KQtpJ:-LIkM93t4QWP7p2KQtpK:b-k7viwz) ([raw file](https://github.com/prestashop/prestashop/blob/b0341caf1bafabd9bf3d50f96ce5226180863c26/admin-dev/themes/default/js/bundle/product/form.js#L755)):*
> ```JavaScript
> 
>           if ($('#form_step2_specific_price_sp_price').val() != '') {
>               $('#form_step2_specific_price_sp_price').prop('disabled', false);
> ```

Please avoid duplicate selectors (store in a variable)
___
*[admin-dev/themes/default/js/bundle/product/form.js, line 775 at r1](https://reviewable.io/reviews/prestashop/prestashop/9388#-LIkLN9L9xi7_32Grjrg:-LIkLN9L9xi7_32Grjrh:b-1hikih) ([raw file](https://github.com/prestashop/prestashop/blob/b0341caf1bafabd9bf3d50f96ce5226180863c26/admin-dev/themes/default/js/bundle/product/form.js#L775)):*
> ```JavaScript
>    * Initialize form behavior
>    */
>   function _init() {
> ```

Why the leading underscore?

https://github.com/airbnb/javascript#naming--leading-underscore
___
*[admin-dev/themes/default/js/bundle/product/form.js, line 821 at r1](https://reviewable.io/reviews/prestashop/prestashop/9388#-LIl-Jke3NN4jSq9QUA3:-LIl-Jkf91p9OQbTFaDK:b-7j9igb) ([raw file](https://github.com/prestashop/prestashop/blob/b0341caf1bafabd9bf3d50f96ce5226180863c26/admin-dev/themes/default/js/bundle/product/form.js#L821)):*
> ```JavaScript
>   }
> 
>   function enableSpecificPriceFieldIfEligible() {
> ```

I like that naming convention :clap: 
___
*[src/PrestaShopBundle/Controller/Admin/SpecificPriceController.php, line 108 at r1](https://reviewable.io/reviews/prestashop/prestashop/9388#-LIl0P8r8x_43TiN431B:-LIl0P8r8x_43TiN431C:btd2qz9) ([raw file](https://github.com/prestashop/prestashop/blob/b0341caf1bafabd9bf3d50f96ce5226180863c26/src/PrestaShopBundle/Controller/Admin/SpecificPriceController.php#L108)):*
> ```PHP
>             } else {
>                 $idSpecificPrice = null;
>             }
> ```

To reduce visual clutter, I think you can use the ternary operator for this one:

```php
$idSpecificPrice = (is_numeric($idSpecificPrice)) ? (int) $idSpecificPrice : null;
```
___
*[src/PrestaShopBundle/Controller/Admin/SpecificPriceController.php, line 130 at r1](https://reviewable.io/reviews/prestashop/prestashop/9388#-LIl1dTT09efJ_fq_oC8:-LIl1dTT09efJ_fq_oC9:b97xfxk) ([raw file](https://github.com/prestashop/prestashop/blob/b0341caf1bafabd9bf3d50f96ce5226180863c26/src/PrestaShopBundle/Controller/Admin/SpecificPriceController.php#L130)):*
> ```
> /ProductPage\Forms/
> ```
Should we use all forward slashes?
___
*[src/PrestaShopBundle/Controller/Admin/SpecificPriceController.php, line 143 at r1](https://reviewable.io/reviews/prestashop/prestashop/9388#-LIl0lRZ5tN4EG2ecZN7:-LIl0lRZ5tN4EG2ecZN8:bwlsent) ([raw file](https://github.com/prestashop/prestashop/blob/b0341caf1bafabd9bf3d50f96ce5226180863c26/src/PrestaShopBundle/Controller/Admin/SpecificPriceController.php#L143)):*
> ```PHP
>             $price = $adminProductWrapper->getSpecificPriceDataById($idSpecificPrice);
>         } catch (PrestaShopObjectNotFoundException $e) {
>             return new Response(sprintf('Cannot find specific price %d', $idSpecificPrice), Response::HTTP_BAD_REQUEST);
> ```

This string must be localized. 

What do you think @louisebonnard ?
___
*[src/PrestaShopBundle/Controller/Admin/SpecificPriceController.php, line 150 at r1](https://reviewable.io/reviews/prestashop/prestashop/9388#-LIl10Hz4AgWnU1c7gMS:-LIl10Hz4AgWnU1c7gMT:ba646sm) ([raw file](https://github.com/prestashop/prestashop/blob/b0341caf1bafabd9bf3d50f96ce5226180863c26/src/PrestaShopBundle/Controller/Admin/SpecificPriceController.php#L150)):*
> ```PHP
>         $options = ['selected_product_attribute' => $price->id_product_attribute];
>         $formBuilder = $this->createFormBuilder();
>         $formBuilder->add('step2', 'PrestaShopBundle\Form\Admin\Product\EmbeddedSpecificPrice', $options);
> ```

I think you can inline $options
___
*[src/PrestaShopBundle/Controller/Admin/SpecificPriceController.php, line 150 at r1](https://reviewable.io/reviews/prestashop/prestashop/9388#-LIl15G1AqsDNTvvZQoH:-LIl15G1AqsDNTvvZQoI:b-kb24y1) ([raw file](https://github.com/prestashop/prestashop/blob/b0341caf1bafabd9bf3d50f96ce5226180863c26/src/PrestaShopBundle/Controller/Admin/SpecificPriceController.php#L150)):*
> ```
> PrestaShopBundle\Form\Admin\Product\EmbeddedSpecificPrice
> ```
Use `EmbeddedSpecificPrice::class` instead
___
*[src/PrestaShopBundle/Controller/Admin/SpecificPriceController.php, line 238 at r1](https://reviewable.io/reviews/prestashop/prestashop/9388#-LIl23qr5SuhfTXCUaVd:-LIl23qr5SuhfTXCUaVe:bneta2y) ([raw file](https://github.com/prestashop/prestashop/blob/b0341caf1bafabd9bf3d50f96ce5226180863c26/src/PrestaShopBundle/Controller/Admin/SpecificPriceController.php#L238)):*
> ```PHP
>      * @throws \PrestaShopDatabaseExceptionCore if date is not valid
>      */
>     private static function formatForDatePicker($dateAsString)
> ```

I don't think this belongs here.

**Comment by LouiseBonnard** [comment]
@eternoendless, @matks, to localize the "Cannot find specific price %d" string, I suggest the _Admin.Catalog.Notification_ domains.

**Comment by matks** [comment]
*Bug found*: https://github.com/PrestaShop/PrestaShop/pull/9388/files#diff-92ead381d47087dcec5699f591e4e4bbR766
triggered too many times, this reloads the page inopportunely

File: admin-dev/themes/default/js/bundle/product/form.js [hunk 2]
```diff
@@ -600,6 +598,7 @@ var specificPrices = (function() {
           elem.addClass('hide');
         }
 
+        // build listing
```
Comment (matks) [inline admin-dev/themes/default/js/bundle/product/form.js]: useless ?

**Comment by matks** [comment]
Closed as too many changes were needed and the ticket scope changed. New PR is here: https://github.com/PrestaShop/PrestaShop/pull/9465

**Pull request closed without merging**